tower-middleware/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+//! Middleware traits and implementations.
+//!
+//! A middleware decorates an service and provides additional functionality.
+//! This additional functionality may include, but is not limited to:
+//!
+//! * Rejecting the request.
+//! * Taking an action based on the request.
+//! * Mutating the request before passing it along to the application.
+//! * Mutating the response returned by the application.
+//!
+//! A middleware implements the [`Middleware`] trait.
+
+#![deny(missing_docs)]
+#![doc(html_root_url = "https://docs.rs/tower-middleware/0.1.0")]
+
+extern crate futures;
+extern crate tower_service;
+
+#[cfg(feature = "ext")]
+mod ext;
+
+#[cfg(feature = "ext")]
+pub use ext::{Chain, MiddlewareExt};
+
+use tower_service::Service;
+
+/// Decorates a `Service`, transforming either the request or the response.
+///
+/// Often, many of the pieces needed for writing network applications can be
+/// reused across multiple services. The `Middleware` trait can be used to write
+/// reusable components that can be applied to very different kinds of services;
+/// for example, it can be applied to services operating on different protocols,
+/// and to both the client and server side of a network transaction.
+///
+/// # Log
+///
+/// Take request logging as an example:
+///
+/// ```rust
+/// # extern crate futures;
+/// # extern crate tower_service;
+/// # extern crate tower_web;
+/// # #[macro_use] extern crate log;
+/// use tower_service::Service;
+/// use tower_web::middleware::Middleware;
```

Comment:
I'm not sure that Tower-web should be a dependency in tower itself. This is a typo, correct?

Reply:
Yup, I fixed the example, let me know if its a bit clearer.